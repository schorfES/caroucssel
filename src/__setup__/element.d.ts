export {};

declare global {
	interface Element {
		mockedTop?: number;
		mockedLeft?: number;
		mockedClientWidth: number;
		mockedOffsetWidth: number;
		mockedClientHeight: number;
		mockedOffsetHeight: number;
	}
}
