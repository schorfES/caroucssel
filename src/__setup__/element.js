function __getSiblings(el) {
	return [...el.parentNode.children];
}


HTMLElement.prototype.mockTop = null;
HTMLElement.prototype.mockLeft = null;
HTMLElement.prototype.mockClientWidth = 120;
HTMLElement.prototype.mockOffsetWidth = 120;
HTMLElement.prototype.mockClientHeight = 50;
HTMLElement.prototype.mockOffsetHeight = 50;

HTMLElement.prototype.scrollTo = function({top, left}) {
	this.mockTop = top;
	this.mockLeft = left;

	const event = document.createEvent('Event');
	event.initEvent('scroll');
	this.dispatchEvent(event);
	jest.runAllTimers();
};

HTMLElement.prototype.getBoundingClientRect = function() {
	const siblings = __getSiblings(this);
	const index = siblings.indexOf(this);
	const width = this.mockClientWidth;
	const height = this.mockClientHeight;
	const top = this.mockTop;
	const right = 0;
	const bottom = 0;
	let left = this.mockLeft;

	if (left === null) {
		left = index * width;
	}

	return {width, height, top, right, bottom, left};
};

Object.defineProperty(HTMLElement.prototype, 'scrollTop', {
	get: function() {
		return this.mockTop;
	}
});

Object.defineProperty(HTMLElement.prototype, 'scrollLeft', {
	get: function() {
		return this.mockLeft;
	}
});

Object.defineProperty(HTMLElement.prototype, 'clientWidth', {
	get: function() {
		return this.mockClientWidth;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetWidth', {
	get: function() {
		return this.mockOffsetWidth;
	}
});

Object.defineProperty(HTMLElement.prototype, 'clientHeight', {
	get: function() {
		return this.mockClientHeight;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetHeight', {
	get: function() {
		return this.mockOffsetHeight;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetTop', {
	get: function() {
		return this.getBoundingClientRect().top;
	}
});

Object.defineProperty(HTMLElement.prototype, 'offsetLeft', {
	get: function() {
		return this.getBoundingClientRect().left;
	}
});
