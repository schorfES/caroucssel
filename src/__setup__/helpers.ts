/* istanbul ignore file */

/**
 * Options of the HTML markup.
 */
export type FixtureOptions = {
	id: string | null;
};

/**
 * Create a fixture of a basic carousel HTML markup with a specific amount of
 * items inside.
 * @internal
 * @param amount the number of items inside this HTML markup
 * @param options options to control the markup
 * @param options.id is the id of the carousel container element
 * @returns the rendered HTML markup
 */
export function fixture(amount: number, options: Partial<FixtureOptions> = {}): string {
	const settings: FixtureOptions = { id: null, ...options };

	return `
		<div class="container">
			<div class="caroucssel"${settings.id ? ` id="${settings.id}"` : ''}>
				${[...Array(amount).keys()].map((index) => `
					<div class="item item-${index}">Item ${index}</div>
				`).join('')}
			</div>
		</div>
	`;
}

/**
 * Triggers a resize event.
 * @internal
 */
export function triggerResize(): void {
	const event = document.createEvent('UIEvents');
	event.initEvent('resize', true, false);
	window.dispatchEvent(event);
	jest.runAllTimers();
}

/**
 * A scroll position.
 */
export type ScrollPosition = {
	left: number;
	top: number;
};

/**
 * Scrolls to a given position and triggers a scroll event on a given element.
 * @internal
 * @param element the element to control.
 * @param position new position of element
 */
export function triggerScroll(element: Element, position: Partial<ScrollPosition>): void {
	element.mockedTop = position.top || 0;
	element.mockedLeft = position.left || 0;

	const event = document.createEvent('Event');
	event.initEvent('scroll');
	element.dispatchEvent(event);
	jest.runAllTimers();
}

/**
 * Triggers a click event on a given element
 * @internal
 * @param element the element to control
 */
export function triggerClick(element: Element): void {
	const event = document.createEvent('Event');
	event.initEvent('click');
	element.dispatchEvent(event);
	jest.runAllTimers();
}

/**
 * Wrapper to ensure the querySelector() returns an element.
 * @internal
 * @param selector the selector passed into querySelector
 * @param root the element root from where to query
 * @returns the found element
 * @throws if no matching element was found
 */
export function querySelector(selector: string, root: Element | Document = document): Element {
	const el = root.querySelector(selector);
	if (!el) {
		throw new Error('Selector not found');
	}

	return el;
}
