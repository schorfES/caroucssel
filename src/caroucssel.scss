@mixin caroucssel() {
	display: flex;
	flex-wrap: nowrap;
	align-items: stretch;
	overflow: auto;
	overflow-y: hidden;
	transform: translate3d(0, 0, 0);
	width: 100%;
	white-space: nowrap;
	scroll-behavior: smooth;
	-ms-overflow-style: -ms-autohiding-scrollbar;
	-webkit-overflow-scrolling: touch;

	> * {
		flex: 0 0 auto;
		outline: none;
		white-space: normal;
	}

	/* browsers with scrollbars that only appear on scrolling */
	/* @FIXME: Is the spaceless variant required? */
	&[style*='margin-bottom: 0px'],
	&[style*='margin-bottom:0px'] {
		scrollbar-width: none;
		-ms-overflow-style: none;

		&::-webkit-scrollbar {
			display: none;
		}
	}
}

@mixin caroucssel-snap($at: 100%) {
	// See: https://medium.com/@OneBitAhead/detecting-css-features-with-supports-1fb9c04c56c0
	scroll-snap-type: mandatory;
	scroll-snap-type: x mandatory; // sass-lint:disable-line no-duplicate-properties
	scroll-snap-destination: $at $at;
	scroll-snap-points-x: snapInterval(0, $at); // sass-lint:disable-line function-name-format
	scroll-snap-points-x: repeat($at); // sass-lint:disable-line no-duplicate-properties

	> * {
		scroll-snap-align: start;
	}
}


%caroucssel {
	@include caroucssel();
}
