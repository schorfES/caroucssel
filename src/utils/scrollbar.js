export class Scrollbar {

	constructor() {
		this._d = null;
		window.addEventListener('resize', () => this._d = null);
	}

	get dimensions() {
		this._d = this._d || (() => {
			const
				inner = document.createElement('div'),
				outer = document.createElement('div')
			;
			let
				width, w1, w2,
				height, h1, h2
			;

			document.body.appendChild(outer);
			outer.style.position = 'absolute';
			outer.style.top = '0px';
			outer.style.left = '0px';
			outer.style.visibility = 'hidden';
			outer.appendChild(inner);

			// Calculate width:
			inner.style.width = '100%';
			inner.style.height = '200px';
			outer.style.width = '200px';
			outer.style.height = '150px';
			outer.style.overflow = 'hidden';
			w1 = inner.offsetWidth;
			outer.style.overflow = 'scroll';
			w2 = inner.offsetWidth;
			w2 = (w1 === w2) ? outer.clientWidth : w2;
			width = w1 - w2;

			// Calculate height:
			inner.style.width = '200px';
			inner.style.height = '100%';
			outer.style.width = '150px';
			outer.style.height = '200px';
			outer.style.overflow = 'hidden';
			h1 = inner.offsetHeight;
			outer.style.overflow = 'scroll';
			h2 = inner.offsetHeight;
			h2 = (h1 === h2) ? outer.clientHeight : h2;
			height = h1 - h2;

			document.body.removeChild(outer);
			return {width, height};
		})();

		return this._d;
	}

}
