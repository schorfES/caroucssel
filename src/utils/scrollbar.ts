import {clearCache, fromCache} from './cache';

export type ScrollbarDimensions = {
	height: number,
};

export class Scrollbar {

	constructor() {
		window.addEventListener('resize', () => {
			clearCache(this, 'dimensions');
		});
	}

	// Inspired by https://gist.github.com/kflorence/3086552
	get dimensions(): ScrollbarDimensions {
		return fromCache<ScrollbarDimensions>(this, 'dimensions', () => {
			const
				inner = document.createElement('div'),
				outer = document.createElement('div')
			;
			let
				// width, w1, w2,
				height, h1, h2
			;

			document.body.appendChild(outer);
			outer.style.position = 'absolute';
			outer.style.top = '0px';
			outer.style.left = '0px';
			outer.style.visibility = 'hidden';
			outer.appendChild(inner);

			// Disabled, not needed for current feature set.
			//
			// Calculate width:
			// inner.style.width = '100%';
			// inner.style.height = '200px';
			// outer.style.width = '200px';
			// outer.style.height = '150px';
			// outer.style.overflow = 'hidden';
			// w1 = inner.offsetWidth;
			// outer.style.overflow = 'scroll';
			// w2 = inner.offsetWidth;
			// w2 = (w1 === w2) ? outer.clientWidth : w2;
			// width = w1 - w2;

			// Calculate height:
			inner.style.width = '200px';
			inner.style.height = '100%';
			outer.style.width = '150px';
			outer.style.height = '200px';
			outer.style.overflow = 'hidden';
			h1 = inner.offsetHeight;
			outer.style.overflow = 'scroll';
			h2 = inner.offsetHeight;
			h2 = (h1 === h2) ? outer.clientHeight : h2;
			height = h1 - h2;

			document.body.removeChild(outer);

			return {
				// width,
				height
			};
		});
	}

}
