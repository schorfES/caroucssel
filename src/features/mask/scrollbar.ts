import { clearCache, fromCache } from '../../utils/cache';


export type ScrollbarDimensions = {
	height: number,
};

/**
 * Helper class for scrollbar features.
 */
export class Scrollbar {

	/**
	 * Creates an instance.
	 * @internal
	 */
	constructor() {
		window.addEventListener('resize', () => {
			clearCache(this, 'dimensions');
		});
	}

	/**
	 * Calculates the dimensions of a scrollbar in the current browser. The result
	 * of the computation will be cached for this instance.
	 *
	 * Inspired by https://gist.github.com/kflorence/3086552
	 *
	 * @return the dimensions of the scrollar
	 */
	get dimensions(): ScrollbarDimensions {
		return fromCache<ScrollbarDimensions>(this, 'dimensions', () => {
			const inner = document.createElement('div');
			const outer = document.createElement('div');

			document.body.appendChild(outer);
			outer.style.position = 'absolute';
			outer.style.top = '0px';
			outer.style.left = '0px';
			outer.style.visibility = 'hidden';
			outer.appendChild(inner);

			// Disabled, not needed for current feature set.
			//
			// Calculate width:
			// inner.style.width = '100%';
			// inner.style.height = '200px';
			// outer.style.width = '200px';
			// outer.style.height = '150px';
			// outer.style.overflow = 'hidden';
			// w1 = inner.offsetWidth;
			// outer.style.overflow = 'scroll';
			// w2 = inner.offsetWidth;
			// w2 = (w1 === w2) ? outer.clientWidth : w2;
			// width = w1 - w2;

			// Calculate height:
			inner.style.width = '200px';
			inner.style.height = '100%';
			outer.style.width = '150px';
			outer.style.height = '200px';
			outer.style.overflow = 'hidden';

			const h1 = inner.offsetHeight;
			outer.style.overflow = 'scroll';
			let h2 = inner.offsetHeight;
			h2 = (h1 === h2) ? outer.clientHeight : h2;
			const height = h1 - h2;

			document.body.removeChild(outer);

			return {
				// width,
				height,
			};
		});
	}

}
