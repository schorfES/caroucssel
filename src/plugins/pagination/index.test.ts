import { fixture, querySelector, triggerClick, triggerResize, triggerScroll } from '../../__setup__/helpers';
import { Carousel } from '../../caroucssel';
import { ScrollbarDimensions } from '../../utils/scrollbar';

import { Pagination, Params, TextParams } from './index';


let mockScrollbarDimensions: ScrollbarDimensions | null = null;

jest.mock('../../utils/scrollbar', () => {
	return {
		Scrollbar: class {
			get dimensions() {
				return mockScrollbarDimensions;
			}
		},
	};
});

describe('Pagination plugin', () => {

	beforeEach(() => {
		jest.useFakeTimers();
		mockScrollbarDimensions = { height: 0 };
	});


	afterEach(() => {
		jest.clearAllTimers();
		document.body.innerHTML = '';
		Carousel.resetInstanceCount();
	});

	it('should add pagination', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');
		new Carousel(el, {
			plugins: [new Pagination()],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
	});

	it('should add pagination with custom options', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		const label: jest.Mock<string, [TextParams]> = jest.fn(
			({ index }) => {
				switch (index) {
					case 0: return 'first';
					case 1: return 'second';
					case 2: return 'third';
					default: return 'un-defined';
				}
			},
		);

		const title: jest.Mock<string, [TextParams]> = jest.fn(
			({ index }) => {
				let name: string;

				switch (index) {
					case 0: name = 'first'; break;
					case 1: name = 'second'; break;
					case 2: name = 'third'; break;
					default: name = 'un-defined'; break;
				}

				return `Go to ${name} item`;
			},
		);

		new Carousel(el, {
			plugins: [
				new Pagination({
					label,
					title,
					className: 'custom-pagination-class',
				}),
			],
		});

		expect(document.body.innerHTML).toMatchSnapshot();

		expect(label).toHaveBeenCalledTimes(3);
		expect(label).toHaveBeenNthCalledWith(1, {
			index: 0,
			page: [0],
			pages: [[0], [1], [2]],
		});

		expect(title).toHaveBeenCalledTimes(3);
		expect(title).toHaveBeenNthCalledWith(3, {
			index: 2,
			page: [2],
			pages: [[0], [1], [2]],
		});
	});

	it('should add pagination with custom template', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		const label: jest.Mock<string, [TextParams]> = jest.fn(
			({ index }) => {
				switch (index) {
					case 0: return 'first';
					case 1: return 'second';
					case 2: return 'third';
					default: return 'un-defined';
				}
			},
		);

		const title: jest.Mock<string, [TextParams]> = jest.fn(
			({ index }) => {
				let name;
				switch (index) {
					case 0: name = 'first'; break;
					case 1: name = 'second'; break;
					case 2: name = 'third'; break;
					default: name = 'un-defined'; break;
				}
				return `Go to ${name} page`;
			},
		);

		const template: jest.Mock<string, [Params]> = jest.fn(
			({ className, controls, pages, label, title }) =>
				`<div class="${className}" aria-controls="${controls}">
					${pages.map((page, index) => `
						<div class="item" aria-label="${title({ index, page, pages })}">
							${label({ index, page, pages })}
						</div>`
					).join('')}
				</div>`,
		);

		new Carousel(el, {
			plugins: [
				new Pagination({
					label,
					title,
					template,
					className: 'custom-pagination-class',
				}),
			],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(1);
	});

	it('should handle pagination with custom template that returns empty string', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		const template: jest.Mock<string, []> = jest.fn(() => '');

		new Carousel(el, {
			plugins: [new Pagination({ template })],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(1);
	});

	it('should handle pagination with custom template that returns null', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		// Test when js custom implementation returns null
		// @ts-ignore
		const template: jest.Mock<string, []> = jest.fn(() => null);

		new Carousel(el, {
			plugins: [new Pagination({ template })],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(1);
	});

	it('should handle pagination with custom template that returns undefined', () => {
		document.body.innerHTML = fixture(3);
		const el = querySelector('.caroucssel');

		// Test when js custom implementation returns undefined
		// @ts-ignore
		const template: jest.Mock<string, []> = jest.fn(() => undefined);

		new Carousel(el, {
			plugins: [new Pagination({ template })],
		});

		expect(document.body.innerHTML).toMatchSnapshot();
		expect(template).toHaveBeenCalledTimes(1);
	});

	it('should not add pagination without items', () => {
		document.body.innerHTML = fixture(0);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 100);
		new Carousel(el, {
			plugins: [new Pagination()],
		});

		const pagination = document.querySelectorAll('.pagination');
		expect(pagination).toHaveLength(0);
	});

	it('should not add pagination with single item', () => {
		document.body.innerHTML = fixture(1);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 100);
		new Carousel(el, {
			plugins: [new Pagination()],
		});

		const pagination = document.querySelectorAll('.pagination');
		expect(pagination).toHaveLength(0);
	});

	it('should re-render pagination on resize', () => {
		document.body.innerHTML = fixture(4);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		new Carousel(el, {
			plugins: [new Pagination()],
		});

		let pagination = document.querySelectorAll('.pagination');
		let pages = pagination[0].querySelectorAll('li');
		expect(pagination).toHaveLength(1);
		expect(pages).toHaveLength(2);

		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 100);
		triggerResize();
		pagination = document.querySelectorAll('.pagination');
		pages = pagination[0].querySelectorAll('li');
		expect(pagination).toHaveLength(1);
		expect(pages).toHaveLength(4);

		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 25);
		triggerResize();
		pagination = document.querySelectorAll('.pagination');
		expect(pagination).toHaveLength(0);

		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 10);
		triggerResize();
		pagination = document.querySelectorAll('.pagination');
		expect(pagination).toHaveLength(0);

		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		triggerResize();
		pagination = document.querySelectorAll('.pagination');
		pages = pagination[0].querySelectorAll('li');
		expect(pagination).toHaveLength(1);
		expect(pages).toHaveLength(2);
	});

	it('should update pagination on scroll', () => {
		document.body.innerHTML = fixture(6);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		new Carousel(el, {
			plugins: [new Pagination()],
		});

		const pagination = document.querySelectorAll<HTMLButtonElement>('.pagination > li > button');
		expect(pagination).toHaveLength(3);
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([true, false, false]);

		triggerScroll(el, { left: 100 });
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([false, true, false]);

		triggerScroll(el, { left: 200 });
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([false, false, true]);
	});

	it('should update pagination manually', () => {
		document.body.innerHTML = fixture(6);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		const carousel = new Carousel(el, {
			plugins: [new Pagination()],
		});

		let pagination = document.querySelectorAll<HTMLButtonElement>('.pagination > li > button');
		expect(pagination).toHaveLength(3);
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([true, false, false]);

		el.scrollTo({ left: 100 });
		carousel.update();
		pagination = document.querySelectorAll<HTMLButtonElement>('.pagination > li > button');
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([false, true, false]);

		el.scrollTo({ left: 200 });
		carousel.update();
		pagination = document.querySelectorAll<HTMLButtonElement>('.pagination > li > button');
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([false, false, true]);
	});

	it('should handle clicks', () => {
		document.body.innerHTML = fixture(6);
		const el = querySelector('.caroucssel');
		el.mockedClientWidth = 100;
		[...document.querySelectorAll('.item')].forEach((item) => item.mockedClientWidth = 50);
		new Carousel(el, {
			plugins: [new Pagination()],
		});

		const callback = jest.spyOn(el, 'scrollTo');
		const pagination = document.querySelectorAll<HTMLButtonElement>('.pagination > li > button');

		triggerClick(pagination[1]);
		expect(callback).toHaveBeenCalledTimes(1);
		expect(callback).toHaveBeenNthCalledWith(1, { left: 100, behavior: 'smooth' });
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([false, true, false]);

		triggerClick(pagination[0]);
		expect(callback).toHaveBeenCalledTimes(2);
		expect(callback).toHaveBeenNthCalledWith(2, { left: 0, behavior: 'smooth' });
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([true, false, false]);

		triggerClick(pagination[2]);
		expect(callback).toHaveBeenCalledTimes(3);
		expect(callback).toHaveBeenNthCalledWith(3, { left: 200, behavior: 'smooth' });
		expect([...pagination].map(({ disabled }) => disabled)).toEqual([false, false, true]);

		triggerClick(pagination[2]); // doesn't work, disabled
		expect(callback).toHaveBeenCalledTimes(3);
	});

});
