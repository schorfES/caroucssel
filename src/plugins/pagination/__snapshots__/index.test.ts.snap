// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination plugin should add pagination 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
  <ul class="pagination">
    <li>
      <button type="button"
              aria-controls="caroucssel-1"
              aria-label="Go to 1. page"
              title="Go to 1. page"
              disabled
      >
        <span>
          1
        </span>
      </button>
    </li>
    <li>
      <button type="button"
              aria-controls="caroucssel-1"
              aria-label="Go to 2. page"
              title="Go to 2. page"
      >
        <span>
          2
        </span>
      </button>
    </li>
    <li>
      <button type="button"
              aria-controls="caroucssel-1"
              aria-label="Go to 3. page"
              title="Go to 3. page"
      >
        <span>
          3
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`Pagination plugin should add pagination with custom options 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
  <ul class="custom-pagination-class">
    <li>
      <button type="button"
              aria-controls="caroucssel-1"
              aria-label="Go to first item"
              title="Go to first item"
              disabled
      >
        <span>
          first
        </span>
      </button>
    </li>
    <li>
      <button type="button"
              aria-controls="caroucssel-1"
              aria-label="Go to second item"
              title="Go to second item"
      >
        <span>
          second
        </span>
      </button>
    </li>
    <li>
      <button type="button"
              aria-controls="caroucssel-1"
              aria-label="Go to third item"
              title="Go to third item"
      >
        <span>
          third
        </span>
      </button>
    </li>
  </ul>
</div>
`;

exports[`Pagination plugin should add pagination with custom template 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
  <div class="custom-pagination-class"
       aria-controls="caroucssel-1"
  >
    <div class="item"
         aria-label="Go to first page"
    >
      first
    </div>
    <div class="item"
         aria-label="Go to second page"
    >
      second
    </div>
    <div class="item"
         aria-label="Go to third page"
    >
      third
    </div>
  </div>
</div>
`;

exports[`Pagination plugin should handle pagination with custom template that returns empty string 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
</div>
`;

exports[`Pagination plugin should handle pagination with custom template that returns null 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
</div>
`;

exports[`Pagination plugin should handle pagination with custom template that returns undefined 1`] = `
<div class="container">
  <div class="caroucssel-mask"
       style="overflow: hidden; height: 100%;"
  >
    <div class="caroucssel"
         id="caroucssel-1"
         style="height: calc(100% + 0px); margin-bottom: 0px;"
    >
      <div class="item item-0">
        Item 0
      </div>
      <div class="item item-1">
        Item 1
      </div>
      <div class="item item-2">
        Item 2
      </div>
    </div>
  </div>
</div>
`;
